feat: rax jsx runtime

7213 opened the pull request:
#### Support jsx-runtime
This module is adapted to react's jsx-runtime module. For more information, please see：[introducing-the-new-jsx-transform](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.htm)

#### usage
```json
{
  "presets": [
    [
      "@babel/preset-react",
      {
        "runtime": "automatic",
        "importSource": "rax/jsx-runtime"
      }
    ]
  ]
}
```

Commit:
feat: jsx-runtime

Commit:
feat: jsx-runtime

Commit:
feat: jsx-runtime

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/alibaba/rax?pullRequest=2101) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/alibaba/rax?pullRequest=2101) before we can accept your contribution.<br/><hr/>**songzheng** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/alibaba/rax?pullRequest=2101) it.</sub>

SoloJiang commented:
这个包直接打在 rax 里就好了吧？

7213 commented:
> 这个包直接打在 rax 里就好了吧？

jsx runtime应该避免使用者对具体jsx库的感知，个人觉得单独的包更合适点。

SoloJiang commented:
> > 这个包直接打在 rax 里就好了吧？
> 
> jsx runtime应该避免使用者对具体jsx库的感知。由编译器自动引入 `import {jsx as _jsx} from "xx-jsx-runtime"` 应该放到单独的包更合适点。

react 现在是怎么做的？

SoloJiang commented:
编译成这样是不是更好：`rax/jsx-runtime`

独立包造成的问题是，runtime 核心能力和 rax 核心包之间可能存在版本差异问题，以及开发者是否感知到这个包的版本

7213 commented:
> 编译成这样是不是更好：`rax/jsx-runtime`
> 
> 独立包造成的问题是，runtime 核心能力和 rax 核心包之间可能存在版本差异问题，以及开发者是否感知到这个包的版本

react是在react package里，我来挪一下

Commit:
feat: jsx-runtime

Commit:
feat: jsx-runtime

Commit:
feat: jsx-runtime

SoloJiang commented:
1. commit 信息可以明确点
2. `jsxs`/`jsx`/`jsxDEV` 应该是不一样的 3 个方法

Commit:
feat: jsx-runtime

This module is adapted to react's jsx-runtime module.
Implement Babel's "automatic" JSX runtime API:
 * 1.jsx(type, props, key)
 * 2.jsxs(type, props, key)
 * 3.jsxDEV(type, props, key, __source, __self)

7213 commented:
> 1. commit 信息可以明确点
> 2. `jsxs`/`jsx`/`jsxDEV` 应该是不一样的 3 个方法

react实现中：jsxs是jsx完全等价的别名哈
jsxDEV babel编译时会自动注入source和self调试信息，`generateVNode` 的实现均已包含

SoloJiang commented:
我指的不是生产模式

SoloJiang commented on packages/rax/src/jsx-runtime/index.js:
```diff
@@ -0,0 +1,50 @@
+      }
+  }
+
+  return {
```
最后返回 RaxElement 是不是更好点

7213 commented:
> 我指的不是生产模式

我来处理，目前是为了体积进行了最小实现，忽略了DEV环境下`jsx`、`jsxs`对DEV下校验部分的代码区别

SoloJiang commented:
可以通过环境变量移除掉 DEV 的代码

Commit:
feat: jsx-runtime

 1.jsx-runtime和createElement组织结构保持一致
 2.区分环境导出不同jsx-runtime，和react一致

Commit:
feat: jsx-runtime

 1.jsx-runtime和createElement组织结构保持一致
 2.区分环境导出不同jsx-runtime，和react一致

Commit:
feat: jsx-runtime

 1.jsx-runtime和createElement组织结构保持一致
 2.区分环境导出不同jsx-runtime，和react一致

SoloJiang commented on packages/rax/jsx-runtime.js:
```diff
@@ -0,0 +1 @@
+export {Fragment, jsx, jsxs, jsxDEV} from './index';
```
+1

SoloJiang commented on packages/rax/src/vdom/element.js:
```diff
@@ -33,10 +33,24 @@ export default function Element(type, key, ref, props, owner) {
+      writable: true,
+      value: __self
+    });
+  }
```
这里抽成一个 utils，避免浪费体积

SoloJiang commented on packages/rax/src/jsx-runtime.js:
```diff
@@ -0,0 +1,105 @@
+  return _jsxWithValidation(type, props, key, true, __source, __self);
+}
+
+const jsx = __DEV__ ? jsxWithValidation : jsxRuntime;
```
`jsxWithValidation` 用额外的文件实现，避免最后生产代码里还遗存

7213 commented on packages/rax/jsx-dev-runtime.js:
```diff
@@ -0,0 +1 @@
+export {Fragment, jsxDEV} from './index';
```
ok,不过`lint:fix`似乎没有自动修正这个style

SoloJiang replied:
仓库有点老，没敢升级新的 lint 规则 =。=

Commit:
feat: jsx-runtime

优化构建体积

SoloJiang commented on packages/rax/src/vdom/jsxValidator.js:
```diff
@@ -0,0 +1,54 @@
+export function jsxWithValidation(type, props, key) {
+  return _jsxWithValidation(type, props, key, false);
+}
+export function jsxsWithValidation(type, props, key) {
```
这里的函数命名可以参考 React，可读性更强一点

SoloJiang reviewed (commented):
整体没啥大的问题了，这两天我来切个 next 分支

Commit:
feat: jsx-runtime

修改jsxValidator导出的命名

Commit:
feat: jsx-runtime

修改jsxValidator导出的命名

7213 commented:
> 整体没啥大的问题了，这两天我来切个 next 分支

ok，到时我逐步提供为Rax App提供一个jsx-runtime的版本

yuchonghua commented:
rax 近期能发一个支持jsx-runtime 的版本吗

yuchonghua commented:
期待有更多rax的文档，而非rax-app的文档

yuchonghua commented:
期待有更多[rax的文档](https://github.com/alibaba/rax/issues/2287) ，而非rax-app的文档

SoloJiang commented:
> rax 近期能发一个支持jsx-runtime 的版本吗

双十二之后发布

yuchonghua commented:
> > rax 近期能发一个支持jsx-runtime 的版本吗
> 
> 双十二之后发布

您好，最近这个有推进吗？已经3月了哦
